emu_sv/sv_backend_impl.py

```diff
@@ -0,0 +1,264 @@
+            )
+        elif isinstance(state, DensityMatrix):
+            return EvolveDensityMatrix.evolve
+        else:
+            raise TypeError(f"Unsupported state type: {type(state)}")
+
+    def _run(self) -> Results:
+        """
+        Runs the simulation.
+
+        Returns:
+            The results of the simulation.
+        """
+        self.time = time.time()
+        state = self.initial_state()
+        state = cast(StateVector, state)
+        stepper = self._choose_stepper(state)
+
+        for step in range(self.nsteps):
+
+            dt = self.target_times[step + 1] - self.target_times[step]
+            state.vector, H = stepper(
+                dt * _TIME_CONVERSION_COEFF,
+                self.omega[step],
+                self.delta[step],
+                self.phi[step],
+                self._pulser_data.full_interaction_matrix,
+                state.vector,
+                self._config.krylov_tolerance,
+            )
+
+            # callbacks in observables and self.statistics in H
+            # have "# type: ignore[arg-type]" because H has it's own type
+            # meaning H is not inherited from Operator class.
+            # We decided that ignore[arg-type] is better compared to
+            # having many unused NotImplemented methods
+            for callback in self._config.observables:
+                callback(
+                    self._config,
+                    self.target_times[step + 1] / self.target_times[-1],
+                    state,
+                    H,  # type: ignore[arg-type]
+                    self.results,
+                )
+
+            self.statistics.data.append(time.time() - self.time)
+            self.statistics(
+                self._config,
+                self.target_times[step + 1] / self.target_times[-1],
+                state,
+                H,  # type: ignore[arg-type]
+                self.results,
+            )
+            self.time = time.time()
+            del H
+
+        return self.results
+
+
+class NoisySVBackendImpl(SVBackendImpl):
+
+    def __init__(self, config: SVConfig, pulser_data: PulserData):
+        """
+        Initializes the NoisySVBackendImpl, master equation version.
+        This class is used to handle the Lindblad operators.
+
+        Args:
+            config: The configuration for the emulator.
+            pulser_data: The data for the sequence to be emulated.
+        """
+
+        super().__init__(config, pulser_data)
+
+        self.pulser_lindblads = pulser_data.lindblad_ops
+
+    def initial_state(self) -> State:
+        if self._config.initial_state is not None:  # fix this with state vector
+            state = self._config.initial_state
+            return DensityMatrix(state.matrix.clone(), gpu=self._config.gpu)
+        else:
+            return DensityMatrix.make(self.nqubits, gpu=self._config.gpu)
+
+    def _run(self) -> Results:
+
+        state = self.initial_state()
```

Comment:
there's a `self.time=time()` missing here. Other than that, the only difference I can see between the two run methods is choosing the stepper and initializing the state. Both these things can be done in the `init` that way you only need a single `run` method.